### PR TITLE
Allow getting and setting CUDA kernel attributes

### LINKED ISSUE #2118: Setting CUDA function attributes
RawKernel is very useful, but currently there appears to be no interface to set the CUDA function attributes.

This is necessary in order to use more than 48KB of shared memory per block in newer GPUs, or to specify the ratio of shared memory to L1 cache (ie, the shared memory "carveout").

Overview of shared memory configuration:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared-memory-7-x

**cudaFuncSetAttribute** documentation:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__HIGHLEVEL.html#group__CUDART__HIGHLEVEL_1g422642bfa0c035a590e4c43ff7c11f8d

### PR DESCRIPTION
**Pull request by leofang**
Takeover of #2120 (which in turn was based on #1874) as @andravin lost interest 😢 

Resolves #2118. 

Most of the credit goes to @grlee77 and @andravin. I took over as we really need the ability of setting shared memory size for CUDA kernels. I'd like to see this capability merged asap if possible.

Changes compared to #2120: 
- The `FunctionAttributes` class is removed. Instead, all function attributes are also `RawKernel`'s attributes now. First, I don't like the idea of having a Python object living in a `cdef`'d class just for accessing attributes. As suggested in https://github.com/cupy/cupy/issues/2118#issuecomment-478311591:
> ...one way to clean up all this attributes stuff without breaking backwards compatibility is to add the "properties" directly to the RawKernel and Device classes. It makes literal sense that function-attributes should be properties of the RawKernel class, and device-attributes should be properties of the Device class.

Second, I don't want to retrieve attributes through a non-standard object. Finally, the current implementation (using `@property` and `@attribute.setter`) is more pythonic, even if it's written in Cython.
- `RawKernel.attributes` now returns a dict of attributes, as suggested by CuPy core dev earlier. This behavior not only avoids the second concern above, but is also in line with `cupy.cuda.device.Device().attributes`. (However, this is read-only, and to set the attributes one needs to access the target attribute of `RawKernel` directly.)
- The tutorial for `RawKernel` is updated
- The test is updated to reflect the change

### CODE REVIEW

**Comment by emcastillo** [comment]
We are solving some CI related issues so the merge will take a bit.

**Comment by grlee77** [comment]
Thanks for updating & reviving this PR @leofang.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
CI has a related failure

```
uilding 'cupy.cuda.driver' extension
11:11:29     x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -D_FORCE_INLINES=1 -I/usr/local/cuda/include -I/usr/include/python3.5m -c cupy/cuda/driver.cpp -o build/temp.linux-x86_64-3.5/cupy/cuda/driver.o
11:11:29     cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
11:11:29     cupy/cuda/driver.cpp: In function 'PyObject* __pyx_f_4cupy_4cuda_6driver_funcSetAttribute(intptr_t, int, int, int)':
11:11:29     cupy/cuda/driver.cpp:4968:133: error: 'cuFuncSetAttribute' was not declared in this scope
11:11:29              __pyx_v_status = cuFuncSetAttribute(((struct CUfunc_st*)__pyx_v_f), ((CUfunction_attribute)__pyx_v_attribute), __pyx_v_value);
11:11:29                                                                                                                                          ^
11:11:29     error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
11:11:29     
```

File: cupy/cuda/cupy_cuda.h [hunk 1]
```diff
@@ -143,6 +144,14 @@ CUresult cuLaunchKernel(...) {
     return CUDA_SUCCESS;
 }
 
+// Function attribute
+CUresult cuFuncGetAttribute(...) {
+    return CUDA_SUCCESS;
+}
+
+CUresult cuFuncSetAttribute(...) {
```
Comment (emcastillo) [inline cupy/cuda/cupy_cuda.h]: Maybe this header is not being included?

**Comment by leofang** [comment]
@emcastillo what's the version of Cython used in that build? I can't reproduce this error locally in a py27 env.

**Comment by leofang** [comment]
I guess it's caused by Cython, not by Python version. It also failed on py35:
https://jenkins.preferred.jp/job/chainer/job/cupy_pr/252/TEST=cupy-doc,label=mn1-p100/console

**Comment by emcastillo** [comment]
Let me confirm it tomorrow!

Sorry and thanks!

**Comment by leofang** [comment]
Sure, no worries!

**Comment by emcastillo** [comment]
should be 0.28.3

**Comment by leofang** [comment]
Tried 0.28.3, worked fine. I am now guessing earlier CUDA versions do not provide `cuFuncSetAttribute`. @emcastillo can you check for me what's the CUDA version in the failing builds?

**Comment by leofang** [comment]
It seems `cuFuncSetAttribute` starts in CUDA v9.0.

**Comment by leofang** [comment]
Please rerun Jenkins for me. Thanks.

**Comment by leofang** [comment]
>> I am now guessing earlier CUDA versions do not provide `cuFuncSetAttribute`. 
>
> It seems cuFuncSetAttribute starts in CUDA v9.0.

I forgot that I can check it myself! Yeah I think this is the reason. From [cupy-py2](https://jenkins.preferred.jp/job/chainer/job/cupy_pr/252/TEST=cupy-py2,label=mn1-p100/console):
```shell
22:11:46     Modules:
22:11:46       cuda      : Yes (version 8000)
22:11:46       cudnn     : Yes (version 5110)
```
From [cupy-doc](https://jenkins.preferred.jp/job/chainer/job/cupy_pr/252/TEST=cupy-doc,label=mn1-p100/console):
```shell
22:11:29     Modules:
22:11:29       cuda      : Yes (version 8000)
22:11:29       cudnn     : Yes (version 6021)
```
Commit 38fcde8 should fix the problem.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
cupy-py3 errors seem irrelevant. cupy-doc should be fixed in 419711a by skipping doctests. The result varies system by system.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Failures are unrelated

**Comment by leofang** [comment]
Thank you, @emcastillo, @andravin, and @grlee77 !

**Pull request merged**